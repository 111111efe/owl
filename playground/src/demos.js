export const HELLO_WORLD = `class HelloWorld extends owl.core.Component {
  inlineTemplate = \`<div>Hello <t t-esc="props.name"/></div>\`;
}

const env = {
    qweb: new owl.core.QWeb()
};

const hello = new HelloWorld(env, { name: "World" });
hello.mount(document.body);
`;
