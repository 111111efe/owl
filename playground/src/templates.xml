<templates>
  <div t-name="tabbed-editor" class="tabbed-editor" t-att-style="props.style">
    <div class="tabBar" t-att-class="{resizeable: props.resizeable}" t-on-mousedown="onMouseDown">
      <a t-if="tabs.js" class="tab" t-att-class="{active: state.currentTab==='js'}" t-on-click="setTab('js')">JS</a>
      <a t-if="tabs.xml" class="tab" t-att-class="{active: state.currentTab==='xml'}" t-on-click="setTab('xml')">XML</a>
      <a t-if="tabs.css" class="tab" t-att-class="{active: state.currentTab==='css'}" t-on-click="setTab('css')">CSS</a>
    </div>
    <div class="code-editor" t-ref="'editor'"></div>
  </div>

  <div t-name="playground" class="playground">
      <div class="left-bar" t-att-style="leftPaneStyle" t-att-class="{split: state.splitLayout}">
        <div class="menubar">
          <a class="btn run-code" t-on-click="runCode">â–¶ Run</a>
          <select t-on-change="setSample">
            <option t-foreach="SAMPLES" t-as="sample" t-key="sample_index">
              <t t-esc="sample.description"/>
            </option>
          </select>
          <div class="layout-selector" t-att-class="{active:state.splitLayout}" t-on-click="toggleLayout">â—«</div>
        </div>
        <t t-if="!state.splitLayout">
          <t t-widget="TabbedEditor" t-props="{js:state.js, css:state.css, xml: state.xml, display: 'js|xml|css'}" t-on-updateCode="updateCode"/>
        </t>
        <t t-else="1">
          <t t-widget="TabbedEditor" t-props="{js:state.js, css:state.css, xml: state.xml, display: 'js', style:topEditorStyle}" t-on-updateCode="updateCode"/>
          <div class="separator horizontal"/>
          <t t-widget="TabbedEditor" t-keepalive="1" t-props="{js:state.js, css:state.css, xml: state.xml, display: 'xml|css', resizeable: true}" t-on-updateCode="updateCode" t-on-updatePanelHeight="updatePanelHeight"/>
        </t>
      </div>
      <div class="separator vertical" t-on-mousedown="onMouseDown"/>
      <div class="right-pane"  t-att-style="rightPaneStyle">
        <div class="welcome" t-if="state.displayWelcome">
          <div>ðŸ¦‰ Odoo Web Library ðŸ¦‰</div>
          <div>v<t t-esc="version"/></div>
          <div class="url"><a href="https://github.com/odoo/owl">https://github.com/odoo/owl</a></div>
          <div class="note">Note: these examples require a recent browser to work without a transpilation step. </div>
        </div>
        <div t-if="state.error" class="error">
          <t t-esc="state.error"/>
        </div>
        <div class="content" t-ref="'content'"/>
      </div>
  </div>
</templates>