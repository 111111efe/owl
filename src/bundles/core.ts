import QWeb from "../core/qweb_vdom";
import * as utils from "../core/utils";
import Widget from "../core/widget";


(<any>window).core = {
    QWeb,
    utils,
    Widget
}