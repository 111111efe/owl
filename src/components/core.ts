import { qweb } from "../qweb/qweb";
import { QWebTemplate } from "../qweb/types";
import { VRootNode } from "../vdom/types";
import { buildTree, patch } from "../vdom/vdom";
import { Component } from "./component";
import { Fiber } from "./fiber";
import { scheduler } from "./scheduler";

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

export interface FComponent<T = any, Env = any> {
  template: string;
  components?: { [key: string]: FComponent<any> | CComponent };
  name?: string;
  setup?: (props: any, env: Env) => T | Promise<T> | void;
}

export type CComponent = typeof Component;

export type CInstance<Props = any, Env = any> = Component<Props, Env>;

export type FInstance<T = {}, Env = any> = T & {
  __owl__: OwlElement<T>;
  props: any;
  env: Env;
};

export interface OwlElement<T = any> {
  fiber: Fiber | null;
  name: string;
  components: { [key: string]: FComponent<T> | CComponent };
  isMounted: boolean;
  instance: FInstance<T> | CInstance | null;
  vnode: VRootNode | null;
  qweb: QWebTemplate;
  isReady: Promise<void> | null;
  children: { [key: string]: OwlElement };
}

// -----------------------------------------------------------------------------
// core object
// -----------------------------------------------------------------------------
export const core = {
  current: null as null | OwlElement,
};

// -----------------------------------------------------------------------------
// Mount
// -----------------------------------------------------------------------------
export type MountTarget = HTMLElement | DocumentFragment;

export interface MountConfig<Env = any> {
  props?: Object;
  env?: Env;
}

export function mount<C extends CComponent>(
  Comp: C,
  target: MountTarget,
  config?: MountConfig
): Promise<InstanceType<C>>;
export function mount(
  comp: CInstance,
  target: MountTarget,
  config?: MountConfig
): Promise<CInstance>;
export function mount<T, Env>(
  Comp: FComponent<T>,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export function mount<T, Env>(
  comp: FInstance,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export async function mount(
  info: any,
  target: MountTarget,
  config: MountConfig = {}
): Promise<CInstance | FInstance> {
  if (!(target instanceof HTMLElement || target instanceof DocumentFragment)) {
    const name =
      info instanceof Component ? info.constructor.name : (info as any).name || "No Name";
    let message = `Component '${name}' cannot be mounted: the target is not a valid DOM node.`;
    message += `\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)`;
    throw new Error(message);
  }
  if (info instanceof Component) {
    // this is a CInstance
    const elem = info.__owl__!;
    return scheduler.addFiber(elem.fiber!).then(() => {
      target.appendChild(info.el!);
      return info;
    });
  }
  let element: OwlElement;
  let props = config.props || {};
  let env = config.env || {};
  if (info.prototype instanceof Component || info === Component) {
    // this is a CComponent (Class Component)
    element = createCComponent(info, props, env);
  } else if ("__owl__" in info) {
    // this is a CInstance or a FInstance
    element = info.__owl__;
  } else {
    // this is a FComponent
    element = createFComponent(info, props, env);
  }
  return scheduler.addFiber(element.fiber!).then(() => {
    const fragment = document.createDocumentFragment();
    buildTree(element.vnode!, (fragment as any) as HTMLElement);
    target.appendChild(fragment);
    if (document.body.contains(target)) {
      mountTree(element);
    }
    return element.instance!;
  });
}

function mountTree(element: OwlElement) {
  for (let k in element.children) {
    mountTree(element.children[k]);
  }
  element.isMounted = true;
  if (element.instance.mounted) {
    element.instance.mounted();
  }
}

// -----------------------------------------------------------------------------
// Component creation
// -----------------------------------------------------------------------------

export function createFComponent(fn: FComponent, props: any, env: any): OwlElement {
  let template: string = fn.template;
  const name = fn.name || "Anonymous Function Component";
  if (!template) {
    throw new Error(`Component "${name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: fn.components || {},
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;

  if (fn.setup) {
    const result = fn.setup(props, env);
    const prom = result instanceof Promise ? result : Promise.resolve(result);
    elem.isReady = prom.then((result) => {
      let instance = result ? Object.create(result) : {};
      instance.props = props;
      instance.__owl__ = elem;
      instance.env = env;
      elem.instance = instance;
    });
  } else {
    const instance: FInstance = { props, env, __owl__: elem };
    elem.instance = instance;
    elem.isReady = Promise.resolve();
  }

  startComponent(elem);
  return elem;
}

export function createCComponent<C extends typeof Component>(
  C: C,
  props: any,
  env: any
): OwlElement {
  let template: string = C.template;
  if (!template) {
    throw new Error(`Component "${C.name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name: C.name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: C.components,
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;
  const component = new C(props, env) as InstanceType<C>;
  component.__owl__ = elem;
  elem.instance = component;
  elem.isReady = component.willStart();
  startComponent(elem);
  return elem;
}

// -----------------------------------------------------------------------------
// starting Components
// -----------------------------------------------------------------------------

async function startComponent(elem: OwlElement) {
  const vnode = elem.qweb.createRoot();
  vnode.hooks.create = (el) => (elem.instance.el = el);
  const fiber = new Fiber(elem, null, vnode);
  fiber.root.counter++;
  elem.fiber = fiber;
  elem.vnode = vnode;
  await elem.isReady!;
  elem.qweb.render(vnode, elem.instance, elem);
  fiber.root.counter--;
}

export function renderToFiber(elem: OwlElement, parentFiber: Fiber | null): Fiber {
  const newRoot = elem.qweb.createRoot();
  const fiber = new Fiber(elem, parentFiber, newRoot);
  elem.fiber = fiber;
  fiber.root.counter++;
  new Promise((resolve) => {
    elem.qweb.render(newRoot, elem.instance!, elem);
    fiber.root.counter--;
    resolve();
  });
  return fiber;
}
// -----------------------------------------------------------------------------
// render
// -----------------------------------------------------------------------------

export function render(elem: OwlElement): Promise<void> {
  let fiber = renderToFiber(elem, null);
  return scheduler.addFiber(fiber).then(() => {
    patch(elem.vnode!, fiber.vnode!);
    while (fiber.next) {
      fiber = fiber.next;
      patch(fiber.elem.vnode!, fiber.vnode!);
    }
  });
}
