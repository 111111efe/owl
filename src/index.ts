import { QWeb } from "./qweb";
import { EventBus } from "./event_bus";
import { Component, PureComponent } from "./component";
import { Store, connect } from "./store";
import * as utils from "./utils";
export const core = {
  QWeb,
  EventBus,
  Component,
  PureComponent,
  Store,
  connect,
  utils
};
