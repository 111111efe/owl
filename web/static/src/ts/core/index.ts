export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
