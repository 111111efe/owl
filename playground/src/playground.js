import { HELLO_WORLD } from "./demos.js";

class Playground extends owl.core.Component {
  inlineTemplate = `
    <div class="playground">
        <div class="left-bar">
            <button t-on-click="runCode">â–¶ Run</button>
            <div class="code-editor" t-ref="editor"></div>
        </div>
        <div class="content">
          <iframe t-ref="iframe"/>
        </div>
    </div>`;

  mounted() {
    this.editor = ace.edit(this.refs.editor);
    this.editor.session.setOption("useWorker", false);
    this.editor.setValue(HELLO_WORLD, -1);
    this.editor.setTheme("ace/theme/monokai");
    this.editor.session.setMode("ace/mode/javascript");
    const doc = this.refs.iframe.contentWindow.document;

    // add owl library
    const owlScript = doc.createElement("script");
    owlScript.src = "libs/owl.js";
    doc.body.appendChild(owlScript);
  }

  runCode() {
    const doc = this.refs.iframe.contentWindow.document;

    // // add owl library
    // const owlScript = doc.createElement("script");
    // owlScript.src = "libs/owl.js";
    // doc.body.appendChild(owlScript);

    const code = this.editor.getValue();
    const script = doc.createElement("script");
    script.type = "text/javascript";
    script.innerHTML = code;
    doc.body.appendChild(script);
    //  var thisDoc = this.contentWindow.document;
    //     if ( ! thisDoc.getElementById(scriptID)) {
    //         var scriptObj = thisDoc.createElement("script");
    //         scriptObj.type = "text/javascript";
    //         scriptObj.id = scriptId;
    //         scriptObj.innerHTML = script;
    //         thisDoc.body.appendChild(scriptObj);
    //     }
    // });
  }
}

document.addEventListener("DOMContentLoaded", async function() {
  const env = {
    qweb: new owl.core.QWeb()
  };

  const playground = new Playground(env);
  playground.mount(document.body);
});
