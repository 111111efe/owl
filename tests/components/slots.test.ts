import { Component, mount, useState, xml } from "../../src/core";
import { fromName, makeTestFixture, nextTick, snapshotTemplateCode } from "../helpers";

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("slots", () => {
  test("simple default slot", async () => {
    class Child extends Component {
      static template = xml`<span><t t-slot="default"/></span>`;
    }

    class Parent extends Component {
      static template = xml`<Child>some text</Child>`;
      static components = { Child };
    }
    await mount(Parent, { target: fixture });

    snapshotTemplateCode(fromName(Parent.template));
    snapshotTemplateCode(fromName(Child.template));
    expect(fixture.innerHTML).toBe("<span>some text</span>");
  });

  test("simple default slot, variation", async () => {
    class Child extends Component {
      static template = xml`<t t-slot="default"/>`;
    }

    class Parent extends Component {
      static template = xml`<Child>some text</Child>`;
      static components = { Child };
    }
    await mount(Parent, { target: fixture });

    snapshotTemplateCode(fromName(Parent.template));
    snapshotTemplateCode(fromName(Child.template));
    expect(fixture.innerHTML).toBe("some text");
  });

  test("fun: two calls to the same slot", async () => {
    class Child extends Component {
      static template = xml`<t t-slot="default"/><t t-slot="default"/>`;
    }

    class Parent extends Component {
      static template = xml`<Child>some text</Child>`;
      static components = { Child };
    }
    await mount(Parent, { target: fixture });

    snapshotTemplateCode(fromName(Parent.template));
    snapshotTemplateCode(fromName(Child.template));
    expect(fixture.innerHTML).toBe("some textsome text");
  });

  test("slot content is bound to caller", async () => {
    class Child extends Component {
      static template = xml`<span><t t-slot="default"/></span>`;
    }

    class Parent extends Component {
      static template = xml`<Child><button t-on-click="inc">some text</button></Child>`;
      static components = { Child };
      state = useState({ value: 0 });
      inc() {
        this.state.value++;
      }
    }
    const parent = await mount(Parent, { target: fixture });

    snapshotTemplateCode(fromName(Parent.template));
    snapshotTemplateCode(fromName(Child.template));
    expect(parent.state.value).toBe(0);
    fixture.querySelector("button")!.click();
    expect(parent.state.value).toBe(1);
  });

  test("can define and call slots", async () => {
    class Dialog extends Component {
      static template = xml`
        <div>
          <div><t t-slot="header"/></div>
          <div><t t-slot="footer"/></div>
        </div>`;
    }

    class Parent extends Component {
      static components = { Dialog };
      static template = xml`
        <div>
          <Dialog>
            <t t-set-slot="header"><span>header</span></t>
            <t t-set-slot="footer"><span>footer</span></t>
          </Dialog>
        </div>`;
    }
    await mount(Parent, { target: fixture });

    snapshotTemplateCode(fromName(Dialog.template));
    snapshotTemplateCode(fromName(Parent.template));
    expect(fixture.innerHTML).toBe(
      "<div><div><div><span>header</span></div><div><span>footer</span></div></div></div>"
    );
  });

  test("no named slot content => just no children", async () => {
    class Dialog extends Component {
      static template = xml`<span><t t-slot="header"/></span>`;
    }
    class Parent extends Component {
      static template = xml`<Dialog/>`;
      static components = { Dialog };
    }
    snapshotTemplateCode(fromName(Dialog.template));
    snapshotTemplateCode(fromName(Parent.template));

    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe("<span></span>");
  });

  test("named slots can define a default content", async () => {
    class Dialog extends Component {
      static template = xml`
          <span>
            <t t-slot="header">default content</t>
          </span>`;
    }
    class Parent extends Component {
      static template = xml`<div><Dialog/></div>`;
      static components = { Dialog };
    }
    snapshotTemplateCode(fromName(Dialog.template));
    snapshotTemplateCode(fromName(Parent.template));

    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe("<div><span>default content</span></div>");
  });

  test("dafault slots can define a default content", async () => {
    class Dialog extends Component {
      static template = xml`
        <span>
          <t t-slot="default">default content</t>
        </span>`;
    }
    class Parent extends Component {
      static template = xml`<div><Dialog/></div>`;
      static components = { Dialog };
    }
    await mount(Parent, { target: fixture });

    snapshotTemplateCode(fromName(Dialog.template));
    expect(fixture.innerHTML).toBe("<div><span>default content</span></div>");
  });

  test("default content is not rendered if slot is provided", async () => {
    class Dialog extends Component {
      static template = xml`
          <span>
            <t t-slot="default">default content</t>
          </span>`;
    }
    class Parent extends Component {
      static template = xml`<div><Dialog>hey</Dialog></div>`;
      static components = { Dialog };
    }
    snapshotTemplateCode(fromName(Dialog.template));
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });

    expect(fixture.innerHTML).toBe("<div><span>hey</span></div>");
  });

  test("default content is not rendered if named slot is provided", async () => {
    class Dialog extends Component {
      static template = xml`
          <span>
            <t t-slot="header">default content</t>
          </span>`;
    }
    class Parent extends Component {
      static template = xml`<div><Dialog><t t-set-slot="header">hey</t></Dialog></div>`;
      static components = { Dialog };
    }
    snapshotTemplateCode(fromName(Dialog.template));
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });

    expect(fixture.innerHTML).toBe("<div><span>hey</span></div>");
  });

  test("slots are rendered with proper context", async () => {
    class Dialog extends Component {
      static template = xml`<span><t t-slot="footer"/></span>`;
    }

    class Parent extends Component {
      static template = xml`
        <div>
          <span class="counter"><t t-esc="state.val"/></span>
          <Dialog>
            <t t-set-slot="footer">
              <button t-on-click="doSomething">do something</button>
            </t>
          </Dialog>
        </div>`;
      static components = { Dialog };
      state = useState({ val: 0 });
      doSomething() {
        this.state.val++;
      }
    }
    snapshotTemplateCode(fromName(Dialog.template));
    snapshotTemplateCode(fromName(Parent.template));

    await mount(Parent, { target: fixture });

    expect(fixture.innerHTML).toBe(
      '<div><span class="counter">0</span><span><button>do something</button></span></div>'
    );

    fixture.querySelector("button")!.click();
    await nextTick();

    expect(fixture.innerHTML).toBe(
      '<div><span class="counter">1</span><span><button>do something</button></span></div>'
    );
  });
});
