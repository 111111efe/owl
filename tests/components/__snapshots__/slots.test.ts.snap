// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slots can define and call slots 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><div><owl-anchor/></div><div><owl-anchor/></div></div>\`);
    children = new Array(2);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'header');
    b1.children[1] = callSlot(ctx, 'footer');
    return b1;
  }
}"
`;

exports[`slots can define and call slots 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  class Block2 extends BNode {
    static el = elem(\`<span>header</span>\`);
  }
  
  class Block3 extends BNode {
    static el = elem(\`<span>footer</span>\`);
  }
  
  const slot3 = ctx => () => {
    const b4 = new Block2();
    return b4;
  }
  
  const slot5 = ctx => () => {
    const b6 = new Block3();
    return b6;
  }
  
  return ctx => {
    const b1 = new Block1();
    const slots2 = {'header': slot3(ctx), 'footer': slot5(ctx)};
    const b7 = b1.children[0] = new BComponent(ctx, \`Dialog\`, {})
    b7.component.__owl__.slots = slots2;
    return b1;
  }
}"
`;

exports[`slots dafault slots can define a default content 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  const defaultSlot2 = ctx => {
    const b3 = new BText(\`default content\`)
    return b3;
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'default', defaultSlot2);
    return b1;
  }
}"
`;

exports[`slots default content is not rendered if named slot is provided 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  const defaultSlot2 = ctx => {
    const b3 = new BText(\`default content\`)
    return b3;
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'header', defaultSlot2);
    return b1;
  }
}"
`;

exports[`slots default content is not rendered if named slot is provided 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  const slot3 = ctx => () => {
    const b4 = new BText(\`hey\`)
    return b4;
  }
  
  return ctx => {
    const b1 = new Block1();
    const slots2 = {'header': slot3(ctx)};
    const b5 = b1.children[0] = new BComponent(ctx, \`Dialog\`, {})
    b5.component.__owl__.slots = slots2;
    return b1;
  }
}"
`;

exports[`slots default content is not rendered if slot is provided 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  const defaultSlot2 = ctx => {
    const b3 = new BText(\`default content\`)
    return b3;
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'default', defaultSlot2);
    return b1;
  }
}"
`;

exports[`slots default content is not rendered if slot is provided 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  const slot3 = ctx => () => {
    const b4 = new BText(\`hey\`)
    return b4;
  }
  
  return ctx => {
    const b1 = new Block1();
    const slots2 = {'default': slot3(ctx)};
    const b5 = b1.children[0] = new BComponent(ctx, \`Dialog\`, {})
    b5.component.__owl__.slots = slots2;
    return b1;
  }
}"
`;

exports[`slots fun: two calls to the same slot 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  const slot2 = ctx => () => {
    const b3 = new BText(\`some text\`)
    return b3;
  }
  
  return ctx => {
    const slots1 = {'default': slot2(ctx)};
    const b4 = new BComponent(ctx, \`Child\`, {})
    b4.component.__owl__.slots = slots1;
    return b4;
  }
}"
`;

exports[`slots fun: two calls to the same slot 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  return ctx => {
    const b1 = new BMulti(2);
    b1.children[0] = callSlot(ctx, 'default');
    b1.children[1] = callSlot(ctx, 'default');
    return b1;
  }
}"
`;

exports[`slots named slots can define a default content 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  const defaultSlot2 = ctx => {
    const b3 = new BText(\`default content\`)
    return b3;
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'header', defaultSlot2);
    return b1;
  }
}"
`;

exports[`slots named slots can define a default content 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = new BComponent(ctx, \`Dialog\`, {})
    return b1;
  }
}"
`;

exports[`slots no named slot content => just no children 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'header');
    return b1;
  }
}"
`;

exports[`slots no named slot content => just no children 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  return ctx => {
    const b1 = new BComponent(ctx, \`Dialog\`, {})
    return b1;
  }
}"
`;

exports[`slots simple default slot 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  const slot2 = ctx => () => {
    const b3 = new BText(\`some text\`)
    return b3;
  }
  
  return ctx => {
    const slots1 = {'default': slot2(ctx)};
    const b4 = new BComponent(ctx, \`Child\`, {})
    b4.component.__owl__.slots = slots1;
    return b4;
  }
}"
`;

exports[`slots simple default slot 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'default');
    return b1;
  }
}"
`;

exports[`slots simple default slot, variation 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  const slot2 = ctx => () => {
    const b3 = new BText(\`some text\`)
    return b3;
  }
  
  return ctx => {
    const slots1 = {'default': slot2(ctx)};
    const b4 = new BComponent(ctx, \`Child\`, {})
    b4.component.__owl__.slots = slots1;
    return b4;
  }
}"
`;

exports[`slots simple default slot, variation 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  return ctx => {
    const b1 = callSlot(ctx, 'default')
    return b1;
  }
}"
`;

exports[`slots slot content is bound to caller 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<button>some text</button>\`);
    handlers = new Array(1);
    build() {
      super.build();
      this.setupHandler(this.el, 0);
    }
  }
  
  const slot2 = ctx => () => {
    const b3 = new Block1();
    b3.handlers[0] = [\`click\`, (e) => owner(ctx)['inc'](e), ctx];
    return b3;
  }
  
  return ctx => {
    ctx = Object.create(ctx);
    ctx[scope] = 1;
    const slots1 = {'default': slot2(ctx)};
    const b4 = new BComponent(ctx, \`Child\`, {})
    b4.component.__owl__.slots = slots1;
    return b4;
  }
}"
`;

exports[`slots slot content is bound to caller 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'default');
    return b1;
  }
}"
`;

exports[`slots slots are rendered with proper context 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<span><owl-anchor/></span>\`);
    children = new Array(1);
  }
  
  return ctx => {
    const b1 = new Block1();
    b1.children[0] = callSlot(ctx, 'footer');
    return b1;
  }
}"
`;

exports[`slots slots are rendered with proper context 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><span class=\\"counter\\"><owl-text/></span><owl-anchor/></div>\`);
    children = new Array(1);
    data = new Array(1);
    update() {
      this.el.firstChild.firstChild.textContent = toString(this.data[0]);
    }
  }
  
  class Block2 extends BNode {
    static el = elem(\`<button>do something</button>\`);
    handlers = new Array(1);
    build() {
      super.build();
      this.setupHandler(this.el, 0);
    }
  }
  
  const slot3 = ctx => () => {
    const b4 = new Block2();
    b4.handlers[0] = [\`click\`, (e) => owner(ctx)['doSomething'](e), ctx];
    return b4;
  }
  
  return ctx => {
    ctx = Object.create(ctx);
    ctx[scope] = 1;
    const b1 = new Block1();
    b1.data[0] = ctx['state'].val;
    const slots2 = {'footer': slot3(ctx)};
    const b5 = b1.children[0] = new BComponent(ctx, \`Dialog\`, {})
    b5.component.__owl__.slots = slots2;
    return b1;
  }
}"
`;
